### ISSUE TITLE
机器学习之线性回归(linear regression)

### ISSUE THREAD

**Issue by towerjoo** [towerjoo/myML#1]
## 版权说明

作者：朱涛
原文地址：https://github.com/towerjoo/myML/issues/1
署名 保持完整性 CC

![linear_regression](https://cloud.githubusercontent.com/assets/145304/23338763/8dd3a43a-fc4e-11e6-97b4-31851f356c4a.png)

## 总的说明

regression，无论是线性回归还是logistic回归，从高层次来看，是我们认为所有的特定组数据是相关的，符合一定的分布规律（可以想象各种曲线），而regression就是来确定这个曲线的参数从而确定这个曲线（函数），继而我们可以根据确定的函数来计算（预测）新的数据的结果。

那么无论是线性回归 (#1) ，通过线性函数来确定这个分布函数；或者logistic回归 ( #2 )，通过p(x)>threshold 来确定分布函数；都是在求得相应的曲线（函数）的参数的过程。

通常的求法便是 yhat - y的距离最短（yhat是我们的函数求得的结果,y是实际的结果），所以通常使用均方差(mean squared error), (yhat-y)^2 ，那么这个求解参数的过程就变成了:

![least_square](https://cloud.githubusercontent.com/assets/145304/23587008/3f31d696-01dd-11e7-895f-5c9fa41554e1.png)

这里的f(x)就是yhat.

或者更一般地，我们认为所有的特定（这是个相对的概念，我们说特定，是指其符合同一分布，但是随着data set变大，模型更加复杂，特定也可以扩展，例如一阶扩展为二阶等）的data set都是符合一定的pattern（也就是上面说的分布，我们用函数可以描述），但是实际情况一般是我们只知道这个特定data set的一部分数据（也就是子集，例如training set, test set），那么限于数据的不完备，必然会存在一定的噪音，所以整个机器学习的过程就是对于不完备的数据（包含有噪音）我们找出其中的pattern（或者叫逼近），那么也包含噪音的处理过程（例如 overfitting #3 )

## 什么是线性回归?

事实上，限于机器学习的年轻（相比于数学，统计学，生物学等），机器学习很多的方法都是来自于其它领域，线性回归也不例外，它是来自于统计学的一个方法。

定义：给定数据集D={(x1, y1), (x2, y2), ... }，我们试图从此数据集中学习得到一个**线性模型**，这个模型尽可能准确地反应x(i)和y(i)的对应关系。这里的线性模型，就是属性(x)的线性组合的函数，可表示为：

![linear_model](https://cloud.githubusercontent.com/assets/145304/23339038/754b1366-fc54-11e6-9edb-ce5a0d5a2b15.png)

向量表示为: 
![linear_vector](https://cloud.githubusercontent.com/assets/145304/23339041/81c127fc-fc54-11e6-8a23-8f38e380b49a.png)

其中，w=(w1; w2;w3; ..., wd) 表示列向量（; 为行分隔，见于matlab/octave中使用）

这里w表示weight，权重的意思，表示对应的属性在预测结果的权重，这个很好理解，权重越大，对于结果的影响越大；更一般化的表示是theta（Andrew ng的课程就是用theta），是线性模型的参数，用于计算结果。

那么通常的线性回归，就变成了如何求得变量参数的问题，根据求得的参数，我们可以对新的输入来计算预测的值。（也可以用于对训练数据计算模型的准确度）

通俗的理解：x(i)就是一个个属性（例如西瓜书中的色泽，根蒂；Andrew ng示例中的房屋面积，卧室数量等），theta(或者w/b)，就是对应属性的参数（或者权重），我们根据已有数据集来求得属性的参数（相当于求得函数的参数），然后根据模型来对于新的输入或者旧的输入来进行预测（或者评估）。

### 解析解法（公式法）

对于数据集D，我们需要根据每组输入(x, y)来计算出线性模型的参数值，那么如何计算？

![cal_linear](https://cloud.githubusercontent.com/assets/145304/23339181/77923e76-fc57-11e6-8805-8fc5ff5ea955.png)

也就是说我们尽量使得f(xi)接近于yi，那么问题来了，我们如何衡量二者的差别？常用的方法是均方误差，也就是（均方误差的几何意义就是欧氏距离，具体的说明可以搜索知乎中的解释）

![least_square](https://cloud.githubusercontent.com/assets/145304/23339206/5c369eaa-fc58-11e6-91eb-299391f72c12.png)

这里的arg min是指后面的表达式值最小时的(w, b)取值。

事实上这就是大名鼎鼎的**最小二乘法** (least square method)，关于这个算法的背后还有发明权争论（撕逼）的过程，有兴趣可以自行搜索阅读。

那么上面的公式我们如何求得参数w,b呢？于是我们又需要一些微积分（calculus)的知识。（插一句：机器学习为什么很多人抱着极大的热情来开始，又很快灰头土脸地离开，大致就是因为这是一个需要较多，不是较深，数学知识的领域，所以不需要灰心，而是赖着性子，我们继续学习和理解）。

这里
![e_linear](https://cloud.githubusercontent.com/assets/145304/23339331/037ce65e-fc5b-11e6-86cf-a82bd40874b7.png)

E是关于(w, b)的凸函数（哈哈，又来了个新的fancy word，简单理解为Andrew ng说的boat shape的函数，也就是有一个最小值），只有一个最小值，而E在最小值时的(w, b)就是我们所要求的参数值，关于凸函数的准确定义可以参考我们大学的高数课程（还记得吗？），周志华西瓜书p54旁注有说明。

对于凸函数E关于w,b导数都为零时，就得到了最优解。

E对w求导：
![dw_linear](https://cloud.githubusercontent.com/assets/145304/23339515/a06050b0-fc5f-11e6-9c50-1e2de23b7120.png)

E对b求导：
![db_linear](https://cloud.githubusercontent.com/assets/145304/23339518/a74b00fa-fc5f-11e6-8ce1-3b35bec201fb.png)

上面我自己进行了推到（当然这是最简单的部分求导，主要就是平方展开），最终令上面的2个导数为0，即可得到w,b的求解公式：

![w_linear](https://cloud.githubusercontent.com/assets/145304/23339580/d928103a-fc60-11e6-88bc-f0f207645304.png)

![b_linear](https://cloud.githubusercontent.com/assets/145304/23339581/de818836-fc60-11e6-8774-a68f78280f0a.png)

其中

![x_ave](https://cloud.githubusercontent.com/assets/145304/23339599/2b07d75a-fc61-11e6-969f-ee75fbfefbb0.png)

是x的平均值。

**以上是对于输入属性为1个的讨论**

对于多个属性的讨论，通常这时就引入了矩阵表示（头又变大，微积分已经早已忘记，线性几何也是啊！），矩阵表示可以很**简洁**地表示出公式，例如 f(xi) = wTxi + b

将b作为w的一个参数（Andrew ng的课程中都是用theta来表示），那么

![multivariate](https://cloud.githubusercontent.com/assets/145304/23339837/a1c2548e-fc65-11e6-97bf-e83244a8ac19.png)

然后对w(hat)求导就可以得到w矩阵的解：

![mv_w](https://cloud.githubusercontent.com/assets/145304/23339846/bcad99e8-fc65-11e6-88dd-ecdf508af0ac.png)

上面这个公式就是Andrew ng给出的解析解（也就是不用梯度下降的算法），当时Andrew ng一笔带过没有给出求解过程，周志华老师p55有简单的求解过程。

这里面存在一个问题就是矩阵的逆是否存在的问题（非满秩矩阵非正定矩阵时，头好大哦，没听过，查查线代的书吧），如果不存在如何处理？Andrew ng提到了2种方法：减少属性个数（d)，正则化（regularization，这个后续有机会可以再写一篇）。这就像解方程，等式个数少于未知数个数时，我们就会解出多组解，上面提到的减少属性个数就是减少未知数个数。

上面说的都是解析解（也就是直接根据公式来计算，相比于下面介绍的逼近法），下面再说说梯度下降（Gradient Descent）大法。

### 梯度下降(Gradient Descent) 逼近法

首先什么是梯度，可以参考下面这张slide（说的严谨而清晰）：

![screen shot 2017-02-26 at 9 10 27 pm](https://cloud.githubusercontent.com/assets/145304/23339972/2e1b109a-fc68-11e6-97ee-797afd0432bd.png)

简单地说，我们需要沿着梯度下降的方向找到一个收敛的参数值，Andrew ng给出的算法如下：

repeat until converge {
  
![gradient-descent](https://cloud.githubusercontent.com/assets/145304/23340010/65f1f0a0-fc69-11e6-8731-4e897a9f46bf.png)

}

其中alpha是learning rate, 是用来控制下降每步的距离（太小收敛会很慢，太大则可能跳过最优点），Andrew ng提到了按照对数的方法来选择，例如0.1, 0.03, 0.01, 0.003, etc.

其中的J函数是所谓的cost function(loss function)，同样在数学、统计学中有大量的应用，是衡量我们的预测函数f(x)精度的函数，同样我们的目标是最小化J。

![cost_function](https://cloud.githubusercontent.com/assets/145304/23340053/7244ca2a-fc6a-11e6-932f-e8c2fc9987c9.png)

注意到其中的参数1/2m，这个参数是可以简化部分求导（消掉2m)。除了参数外，其它部分与解析解部分是完全相同的（事实上，解析解部分的公式也是cost function)。
这里的参数是不重要的，因为alpha的选择可以对冲掉。

这个过程可以写程序来完成，具体可以做下Andrew ng的相应部分练习题，完成的过程还是很有意思的，也是不断加深对于梯度下降方法的理解。

注意上面的theta和解析解部分的w,b是一样的（只是换了下字母）。

### 如何选择

那么两种方法（解析解和逼近法）如何选择，解析解好处是显而易见的，例如无需选择learning rate alpha，无需循环，坏处是要有复杂的矩阵运算（转置矩阵，逆矩阵等），复杂度比较高O(n^3)，相较而言，梯度下降则是O(d*n^2) 

Andrew ng提到n<10000，大致可以用解析解，n>10000则可以用梯度下降。当然这不是必然的，我们需要理解为什么不选择解析解，主要是我们的计算机太慢（复杂度太高导致时间过长），随着计算机越来越快，这个经验值可以调整。

## 为什么进行线性回归?

周志华老师提到的西瓜判别，Andrew ng提到的房价预测，都是很好的应用。线性回归鉴于其简单易用，易理解（注意上面提到的权重的理解），所以得到了很广泛的应用，在某些场景下或许有优于其它复杂方法的表现。

另外，随着机器学习的发展，近期出现的explainable AI(也就是不只要预测出结果，还有给出预测的原因解释),而线性回归方法则很符合explainable的要求。（另外决策树也很explainable)

## 最后

Andrew ng在讲完线性回归后突然发出了如下的感慨：

> 95%的硅谷做AI的engineer都不懂线性回归，而当你懂了线性回归后，你就超过了大多数硅谷搞AI的人，可以去申请Offer了。

很夸张甚至励志是吧？但是你要记得他录视频是在2011年10月左右（source: 演示的电脑)，6年过去了，懂个线性回归显然不够了，所以我们还是要继续学习其它的ML知识，譬如下一篇对数几率回归(logistic regression #2 )

## 参考资料

1. 周志华 《机器学习》
2. Andrew ng: Machine Learning in Coursera
3. https://en.wikipedia.org/wiki/Linear_regression
4. http://latex2png.com/ latex转换为图片（markdown不支持公式）
5. http://songshuhui.net/archives/76501
6. https://en.wikipedia.org/wiki/Gradient_descent
7. http://math.fudan.edu.cn/gdsx/KEJIAN/方向导数和梯度.pdf
8. https://en.wikipedia.org/wiki/Loss_function
9. http://www.darpa.mil/program/explainable-artificial-intelligence

**Comment by aaaron7** [comment]
给定数据集D={(x1, y1), (x2, y2), ... }，我们试图从此数据集中学习得到一个线性模型，

这里感觉数据集应该是是 D = {(x11;x12;x13.. x1d, y1), (x21;x22;x23...x2d, y2)....} 这样表示更准确一些吧。

**Comment by towerjoo** [comment]
@aaaron7 哈哈，谁知道我的x_i不是个vector?

**Comment by aaaron7** [comment]
没毛病。老铁
